system-parachains/asset-hubs/asset-hub-kusama/src/verify_xcm_upgrade_migration.rs

```diff
@@ -0,0 +1,223 @@
+		>;
+	}
+
+	/// Test migration to verify XCM V4 to V5 compatibility for ForeignAssets and AssetConversion
+	/// storage. This migration doesn't actually alter storage, it only verifies that:
+	/// 1. XCM V4 encoded locations can be decoded as V5
+	/// 2. Location encoding remains the same between V4 and V5
+	pub struct TestXcmV4ToV5Compatibility<T>(core::marker::PhantomData<T>);
+
+	impl<T> OnRuntimeUpgrade for TestXcmV4ToV5Compatibility<T>
+	where
+		T: frame_system::Config
+			+ pallet_assets::Config<ForeignAssetsInstance>
+			+ pallet_asset_conversion::Config<
+				PoolId = (
+					<T as pallet_asset_conversion::Config>::AssetKind,
+					<T as pallet_asset_conversion::Config>::AssetKind,
+				),
+			>,
+		T::AssetKind: From<xcm::v5::Location> + Into<xcm::v5::Location>,
+		T::PoolId: Into<(T::AssetKind, T::AssetKind)>,
+		<T as pallet_assets::Config<ForeignAssetsInstance>>::AssetId:
+			From<xcm::v5::Location> + Into<xcm::v5::Location>,
+	{
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<Vec<u8>, TryRuntimeError> {
+			log::info!(target: LOG_TARGET, "XCM V4 to V5 upgrade compatibility test starting");
+
+			log::info!("Starting XCM V4 to V5 compatibility test migration");
+
+			// Test ForeignAssets storage items
+			Self::test_foreign_assets_compatibility();
+
+			// Test AssetConversion storage items
+			Self::test_asset_conversion_compatibility();
+
+			log::info!("XCM V4 to V5 compatibility test migration completed successfully");
+
+			Ok(Vec::new())
+		}
+
+		#[cfg(feature = "try-runtime")]
+		fn post_upgrade(_state: Vec<u8>) -> Result<(), TryRuntimeError> {
+			// Test a few sample conversions to ensure compatibility
+			Self::test_sample_location_conversions()?;
+
+			log::info!(target: LOG_TARGET, "XCM V4 to V5 compatibility test migration validated successfully");
+			Ok(())
+		}
+	}
+
+	impl<T> TestXcmV4ToV5Compatibility<T>
+	where
+		T: frame_system::Config
+			+ pallet_assets::Config<ForeignAssetsInstance>
+			+ pallet_asset_conversion::Config
+			+ pallet_asset_conversion::Config<
+				PoolId = (
+					<T as pallet_asset_conversion::Config>::AssetKind,
+					<T as pallet_asset_conversion::Config>::AssetKind,
+				),
+			>,
+		T::PoolId: Into<(T::AssetKind, T::AssetKind)>,
+		T::AssetKind: From<xcm::v5::Location> + Into<xcm::v5::Location>,
+		<T as pallet_assets::Config<crate::ForeignAssetsInstance>>::AssetId:
+			From<xcm::v5::Location> + Into<xcm::v5::Location>,
+	{
+		#[cfg(feature = "try-runtime")]
+		fn test_foreign_assets_compatibility() -> Weight {
+			let weight = T::DbWeight::get().reads(1);
+			let tested_assets = 0u32;
+
+			log::info!(target: LOG_TARGET, "Found XCM V4 asset keys: {}", asset_as_v4::Asset::<T>::iter_keys().count());
+			log::info!(target: LOG_TARGET, "Found XCM V5 asset keys: {}", asset_as_v5::Asset::<T>::iter_keys().count());
+
+			assert!(asset_as_v4::Asset::<T>::iter_keys().eq(asset_as_v5::Asset::<T>::iter_keys()));
```

Comment:
@claravanstaden when this assert fails, we will see just true/false, but I would be pretty specific and exact here, if something is not eq, let's log exactly which key does not match and so on, maybe do some zip/enumerate comaprision - something like this **pseudocode:**
```
for (idx, (a, b)) in iter1.into_iter().zip(iter2).enumerate() {
        // TODO: change to `ensure!`
        if a != b {
            println!("Difference at index {}: left = {:?}, right = {:?}", idx, a, b);
            return;
        }
    }
```

This relates to the comment above about logging the count :)

Reply:
Good idea! Done in https://github.com/polkadot-fellows/runtimes/pull/753/commits/1f88a47b7d3a6d16a4be5607fa4935cde144ed24.